File: app/src/main/java/com/absoluteapps/arthurl/mobooru/DataAdapter.java
package com.absoluteapps.arthurl.mobooru;

import android.app.Activity;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.etsy.android.grid.util.DynamicHeightImageView;
import com.squareup.picasso.Picasso;

import java.util.List;

public class DataAdapter extends ArrayAdapter<Data> {

    Activity activity;
    int resource;
    List<Data> datas;
    Boolean showNsfw;
    Uri nsfwLogo = Uri.parse("drawable/nsfwlogo.png");

    public DataAdapter(Activity activity, int resource, List<Data> objects, Boolean shwNsfw) {
        super(activity, resource, objects);

        this.activity = activity;
        this.resource = resource;
        this.datas = objects;
        this.showNsfw = shwNsfw;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View row = convertView;
        final DealHolder holder;

        if (row == null) {
            LayoutInflater inflater = activity.getLayoutInflater();
            row = inflater.inflate(resource, parent, false);

            holder = new DealHolder();
            holder.image = (DynamicHeightImageView) row.findViewById(R.id.image);
            holder.title = (TextView) row.findViewById(R.id.title);
            holder.description = (TextView) row.findViewById(R.id.description);
            holder.score = (TextView) row.findViewById(R.id.score);

            row.setTag(holder);
        } else {
            holder = (DealHolder) row.getTag();
        }

        final Data data = datas.get(position);

        holder.title.setText(data.title.length() > 35 ? data.title.substring(0, 30) + "..." : data.title);
        holder.score.setText(data.score + "🔺");
        if (data.ogSrc == null) {
            holder.description.setText("");
        } else {
            holder.description.setText(data.series.length() > 30 ? data.series.substring(0, 25) + "..." : data.series);
        }
        if (data.thumbImgUrl != ""){
            if (data.nsfw && !showNsfw) {
                holder.image.setHeightRatio(1);
                Picasso.get()
                        .load(R.drawable.nsfwlogo)
                        .resize(500,580)
                        .into(holder.image);
            } else {
                double r = data.rat < 1 ? 1/data.rat : data.rat;
                holder.image.setHeightRatio(r);
                Picasso.get()
                        .load(data.thumbImgUrl)
                        .resize(540, (int) Math.floor(540*r))   // set height programmatically
                        .centerCrop()
                        .transform(new RoundedTransformation(20, 10))
                        .into(holder.image);

            }
        } else {
//            Picasso.with(this.getContext())
//                    .load(new File("drawable/404_notfound.jpg"))
//                    .transform(new RoundedTransformation(20, 10))
//                    .into(holder.image);
        }
        return row;
    }

    static class DealHolder {
        DynamicHeightImageView image;
        TextView title;
        TextView score;
        TextView description;
    }
}

File: app/src/main/java/com/absoluteapps/arthurl/mobooru/DataAdapter.kt
package com.absoluteapps.arthurl.mobooru

import android.app.Activity
import android.graphics.Typeface
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.MultiTransformation
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions

import com.etsy.android.grid.util.DynamicHeightImageView
import jp.wasabeef.glide.transformations.BlurTransformation
import jp.wasabeef.glide.transformations.RoundedCornersTransformation

class DataAdapter(internal var activity: Activity, internal var resource: Int, internal var datas: List<Data>, internal var showNsfw: Boolean = false, internal var showTitles: Boolean = true) : ArrayAdapter<Data>(activity, resource, datas) {
    private val regOpts = RequestOptions()
            .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
            .transform(MultiTransformation(CenterCrop(), RoundedCornersTransformation(20,10)))

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var row = convertView
        val holder: DealHolder

        if (row == null) {
            val inflater = activity.layoutInflater
            row = inflater.inflate(resource, parent, false)

            holder = DealHolder()
            holder.image = row!!.findViewById<View>(R.id.image) as DynamicHeightImageView
            holder.title = row.findViewById<View>(R.id.title) as TextView
            holder.title!!.visibility = if (showTitles) View.VISIBLE else View.GONE
            holder.description = row.findViewById<View>(R.id.description) as TextView
            holder.description!!.visibility = if (showTitles) View.VISIBLE else View.GONE
            holder.score = row.findViewById<View>(R.id.score) as TextView
            holder.score!!.visibility = if (showTitles) View.VISIBLE else View.GONE

            row.tag = holder
        } else {
            holder = row.tag as DealHolder
        }

        val data = datas[position]

        holder.title!!.text = if (data.title.length > 35) data.title.substring(0, 30) + "..." else data.title
        holder.score!!.text = data.score
        holder.score!!.setTypeface(null, Typeface.BOLD_ITALIC)
        holder.description!!.text = if (data.series.length > 30) data.series.substring(0, 25) + "..." else data.series

        if (data.thumbImgUrl !== "") {
            val r = if (data.rat < 1) 1 / data.rat else data.rat

            if (data.nsfw && !showNsfw) {
                holder.image!!.heightRatio = r
                Glide.with(activity)
                        .load(data.thumbImgUrl)
                        .override(540, Math.floor(540 * r).toInt())   // set height programmatically
                        .apply(RequestOptions().transform(MultiTransformation(BlurTransformation(25, 3), CenterCrop(), RoundedCornersTransformation(20,10))))
                        .transition(DrawableTransitionOptions.withCrossFade())
                        .into(holder.image!!)
            } else {
                holder.image!!.heightRatio = r
                Glide.with(activity)
                        .load(data.thumbImgUrl)
                        .override(540, Math.floor(540 * r).toInt())   // set height programmatically
                        .apply(regOpts)
                        .into(holder.image!!)
            }
        }
        return row
    }

    internal class DealHolder {
        var image: DynamicHeightImageView? = null
        var title: TextView? = null
        var score: TextView? = null
        var description: TextView? = null
    }
}
